core/src/main/java/com/linecorp/armeria/client/circuitbreaker/AbstractCircuitBreakerClient.java

```diff
@@ -115,13 +124,24 @@ final CircuitBreakerRule fromRuleWithContent() {
         return fromRuleWithContent;
     }
 
+    final CircuitBreakerClientHandler handler() {
+        return handler;
+    }
+
     @Override
     public final O execute(ClientRequestContext ctx, I req) throws Exception {
-        final CircuitBreakerCallback callback = handler.tryRequest(ctx, req);
-        if (callback == null) {
-            return unwrap().execute(ctx, req);
+        try {
+            final CircuitBreakerCallback callback = handler.tryRequest(ctx, req);
+            if (callback == null) {
+                return unwrap().execute(ctx, req);
+            }
+            return doExecute(ctx, req, callback);
+        } catch (Exception ex) {
+            if (handler().isCircuitBreakerException(ex) && fallback != null) {
+                return fallback.apply(ctx, req);
```

Comment:
It should be better to early check nullness. 
```suggestion
                final O res = fallback.apply(ctx, req);
                return requireNonNull(res, "fallback.apply() returned null");
```